File: api/internal/http/analytics.go
package http

import (
	"net/http"
	"strconv"
	"time"

	"decorebator.com/internal/common"
	"decorebator.com/internal/model"
	"decorebator.com/internal/service"
	"github.com/gin-gonic/gin"
)

// getUserFromContext extracts the user object from the gin context
func getUserFromContext(c *gin.Context) (*model.User, bool) {
	user, exists := c.Get("user")
	if !exists {
		return nil, false
	}
	userObj, ok := user.(*model.User)
	return userObj, ok
}

// getCacheTTL returns the cache TTL based on user's subscription
func getCacheTTL(subscriptionPlan model.SubscriptionPlan) time.Duration {
	if subscriptionPlan == model.PlanFree {
		return 1 * time.Hour
	}
	return 1 * time.Minute
}

// RegisterAnalyticsRoutes registers all analytics endpoints
func RegisterAnalyticsRoutes(r *gin.RouterGroup) {
	analytics := r.Group("/analytics")

	analytics.GET("/wordlists/:id/stats", getWordlistStats)
	analytics.GET("/wordlists/:id/word-mastery", getWordlistWordMastery)
	analytics.GET("/wordlists/:id/learning-progress", getWordlistLearningProgress)
	analytics.GET("/wordlists/:id/practice-time", getWordlistPracticeTime)
	analytics.GET("/wordlists/:id/box-distribution-history", getWordlistBoxDistributionHistory)
	analytics.GET("/wordlists/:id/current-box-distribution", getWordlistCurrentBoxDistribution)
	analytics.GET("/wordlists/:id/quiz-type-performance", getWordlistQuizTypePerformance)
	analytics.GET("/progress-summary", getProgressSummary)
}

// getWordlistWordMastery returns word mastery statistics for a wordlist
func getWordlistWordMastery(c *gin.Context) {
	wordlistID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid wordlist ID"})
		return
	}

	userID := c.GetInt64("userID")
	userObj, ok := getUserFromContext(c)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User not found in context"})
		return
	}

	// Verify wordlist ownership
	wordlist, err := service.GetWordlistById(wordlistID, userID)
	if err != nil || wordlist == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Wordlist not found"})
		return
	}

	analyticsService, err := service.NewAnalyticsService(service.AnalyticsConfig{
		UserID:     userID,
		WordlistID: wordlistID,
		UseCache:   true,
		CacheTTL:   getCacheTTL(userObj.SubscriptionPlan),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initialize analytics"})
		return
	}

	stats, err := analyticsService.WordMastery(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch mastery stats"})
		return
	}

	// Ensure stats is never nil - return empty array if no data
	if stats == nil {
		stats = []model.WordMasteryStats{}
	}

	c.JSON(http.StatusOK, gin.H{
		"wordlist_id": wordlistID,
		"stats":       stats,
	})
}

// getWordlistLearningProgress returns daily learning progress
func getWordlistLearningProgress(c *gin.Context) {
	wordlistID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid wordlist ID"})
		return
	}

	// Get days parameter (default 30)
	days := 30
	if d := c.Query("days"); d != "" {
		if parsed, err := strconv.Atoi(d); err == nil && parsed > 0 && parsed <= 365 {
			days = parsed
		}
	}

	userID := c.GetInt64("userID")

	// Verify wordlist ownership
	wordlist, err := service.GetWordlistById(wordlistID, userID)
	if err != nil || wordlist == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Wordlist not found"})
		return
	}

	userObj, ok := getUserFromContext(c)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User not found in context"})
		return
	}

	analyticsService, err := service.NewAnalyticsService(service.AnalyticsConfig{
		UserID:     userID,
		WordlistID: wordlistID,
		UseCache:   true,
		CacheTTL:   getCacheTTL(userObj.SubscriptionPlan),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initialize analytics"})
		return
	}

	progress, err := analyticsService.LearningProgress(c.Request.Context(), days)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch progress"})
		return
	}

	// Ensure progress is never nil - return empty array if no data
	if progress == nil {
		progress = []model.LearningProgressStats{}
	}

	c.JSON(http.StatusOK, gin.H{
		"wordlist_id": wordlistID,
		"days":        days,
		"progress":    progress,
	})
}

// getWordlistBoxDistributionHistory returns historical box distribution
func getWordlistBoxDistributionHistory(c *gin.Context) {
	wordlistID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid wordlist ID"})
		return
	}

	// Get days parameter (default 30)
	days := 30
	if d := c.Query("days"); d != "" {
		if parsed, err := strconv.Atoi(d); err == nil && parsed > 0 && parsed <= 365 {
			days = parsed
		}
	}

	userID := c.GetInt64("userID")

	// Verify wordlist ownership
	wordlist, err := service.GetWordlistById(wordlistID, userID)
	if err != nil || wordlist == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Wordlist not found"})
		return
	}

	userObj, ok := getUserFromContext(c)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User not found in context"})
		return
	}

	analyticsService, err := service.NewAnalyticsService(service.AnalyticsConfig{
		UserID:     userID,
		WordlistID: wordlistID,
		UseCache:   true,
		CacheTTL:   getCacheTTL(userObj.SubscriptionPlan),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initialize analytics"})
		return
	}

	distribution, err := analyticsService.BoxDistributionHistory(c.Request.Context(), days)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch distribution"})
		return
	}

	// Ensure distribution is never nil - return empty array if no data
	if distribution == nil {
		distribution = []map[string]interface{}{}
	}

	c.JSON(http.StatusOK, gin.H{
		"wordlistId":   wordlistID,
		"days":         days,
		"distribution": distribution,
	})
}

// getWordlistQuizTypePerformance returns performance statistics by quiz type for a specific wordlist
func getWordlistQuizTypePerformance(c *gin.Context) {
	wordlistID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid wordlist ID"})
		return
	}

	userID := c.GetInt64("userID")

	// Verify wordlist ownership
	wordlist, err := service.GetWordlistById(wordlistID, userID)
	if err != nil || wordlist == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Wordlist not found"})
		return
	}

	userObj, ok := getUserFromContext(c)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User not found in context"})
		return
	}

	analyticsService, err := service.NewAnalyticsService(service.AnalyticsConfig{
		UserID:     userID,
		WordlistID: wordlistID,
		UseCache:   true,
		CacheTTL:   getCacheTTL(userObj.SubscriptionPlan),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initialize analytics"})
		return
	}

	performance, err := analyticsService.QuizTypePerformance(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch performance stats"})
		return
	}

	// Ensure performance is never nil - return empty array if no data
	if performance == nil {
		performance = []model.QuizTypePerformance{}
	}

	c.JSON(http.StatusOK, gin.H{
		"wordlistId":      wordlistID,
		"quizPerformance": performance,
	})
}

// getWordlistCurrentBoxDistribution returns current distribution of words across Leitner boxes
func getWordlistCurrentBoxDistribution(c *gin.Context) {
	wordlistID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid wordlist ID"})
		return
	}

	userID := c.GetInt64("userID")

	// Verify wordlist ownership
	wordlist, err := service.GetWordlistById(wordlistID, userID)
	if err != nil || wordlist == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Wordlist not found"})
		return
	}

	userObj, ok := getUserFromContext(c)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User not found in context"})
		return
	}

	analyticsService, err := service.NewAnalyticsService(service.AnalyticsConfig{
		UserID:     userID,
		WordlistID: wordlistID,
		UseCache:   true,
		CacheTTL:   getCacheTTL(userObj.SubscriptionPlan),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initialize analytics"})
		return
	}

	distribution, err := analyticsService.CurrentBoxDistribution(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch distribution"})
		return
	}

	// Ensure distribution is never nil - return empty distribution if no data
	if distribution == nil {
		distribution = &model.BoxDistribution{
			Box1Count: 0, Box2Count: 0, Box3Count: 0, Box4Count: 0,
			Box5Count: 0, Box6Count: 0, Box7Count: 0, TotalWords: 0,
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"wordlistId":   wordlistID,
		"distribution": distribution,
		"totalWords":   distribution.TotalWords,
	})
}

// getWordlistPracticeTime returns daily practice time statistics for a wordlist
func getWordlistPracticeTime(c *gin.Context) {
	wordlistID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid wordlist ID"})
		return
	}

	// Get days parameter (default 7)
	days := 7
	if d := c.Query("days"); d != "" {
		if parsed, err := strconv.Atoi(d); err == nil && parsed > 0 && parsed <= 30 {
			days = parsed
		}
	}

	userID := c.GetInt64("userID")

	// Verify wordlist ownership
	wordlist, err := service.GetWordlistById(wordlistID, userID)
	if err != nil || wordlist == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Wordlist not found"})
		return
	}

	userObj, ok := getUserFromContext(c)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User not found in context"})
		return
	}

	analyticsService, err := service.NewAnalyticsService(service.AnalyticsConfig{
		UserID:     userID,
		WordlistID: wordlistID,
		UseCache:   true,
		CacheTTL:   getCacheTTL(userObj.SubscriptionPlan),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initialize analytics"})
		return
	}

	practiceTime, err := analyticsService.PracticeTime(c.Request.Context(), days)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch practice time"})
		return
	}

	// Ensure practiceTime is never nil - return empty array if no data
	if practiceTime == nil {
		practiceTime = []model.PracticeTimeStats{}
	}

	c.JSON(http.StatusOK, gin.H{
		"wordlistId":   wordlistID,
		"days":         days,
		"practiceTime": practiceTime,
	})
}

// getWordlistStats returns statistics for a specific wordlist
func getWordlistStats(c *gin.Context) {
	wordlistID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid wordlist ID"})
		return
	}

	userID := c.GetInt64("userID")

	// Verify wordlist ownership
	wordlist, err := service.GetWordlistById(wordlistID, userID)
	if err != nil || wordlist == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Wordlist not found"})
		return
	}

	userObj, ok := getUserFromContext(c)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User not found in context"})
		return
	}

	analyticsService, err := service.NewAnalyticsService(service.AnalyticsConfig{
		UserID:     userID,
		WordlistID: wordlistID,
		UseCache:   true,
		CacheTTL:   getCacheTTL(userObj.SubscriptionPlan),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initialize analytics"})
		return
	}

	stats, err := analyticsService.Stats(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch wordlist stats"})
		return
	}

	// Ensure stats is never nil - return safe defaults if no data
	if stats == nil {
		stats = &model.WordlistStats{
			TotalWords: 0, WordsMastered: 0, AverageMastery: nil,
			BestStreak: nil, WordsStudiedToday: 0, QuizzesToday: 0,
			AccuracyToday: nil, CurrentStreak: 0,
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"wordlistId": wordlistID,
		"stats":      stats,
	})
}

// getProgressSummary returns progress summary for all user's wordlists
func getProgressSummary(c *gin.Context) {
	userID := c.GetInt64("userID")

	userObj, ok := getUserFromContext(c)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User not found in context"})
		return
	}

	analyticsService, err := service.NewAnalyticsService(service.AnalyticsConfig{
		UserID:     userID,
		WordlistID: 0, // Not needed for progress summary
		UseCache:   true,
		CacheTTL:   getCacheTTL(userObj.SubscriptionPlan),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initialize analytics"})
		return
	}

	summary, err := analyticsService.ProgressSummary(c.Request.Context())
	if err != nil {
		common.Logger.Error("Failed to fetch progress summary", "error", err, "userID", userID)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch progress summary"})
		return
	}

	// Ensure summary is never nil and has valid wordlists array
	if summary == nil {
		summary = &model.ProgressSummaryResponse{Wordlists: []model.WordlistProgress{}}
	} else if summary.Wordlists == nil {
		summary.Wordlists = []model.WordlistProgress{}
	}

	c.JSON(http.StatusOK, summary)
}

File: mobile/api/analytics.ts
import { callAPI } from "./api";
import { getApiBaseUrl } from "./baseUrl";
import { getAllPages, PaginationError } from "./pagination";

export interface WordlistStats {
  totalWords: number;
  wordsMastered: number;
  averageMastery: number;
  bestStreak: number;
  currentStreak: number;
  wordsStudiedToday: number;
  quizzesToday: number;
  accuracyToday: number;
}

export interface WordMasteryStats {
  wordId: number;
  word: string;
  masteryLevel: number;
  accuracy: number;
  streakCount: number;
  highestBox: number;
}

export interface LearningProgress {
  date: string;
  wordsStudied: number;
  accuracyRate: number;
}

export interface QuizTypePerformance {
  quizType: string;
  successRate: number;
  totalAttempts: number;
}

export interface BoxDistribution {
  box1: number;
  box2: number;
  box3: number;
  box4: number;
  box5: number;
  box6: number;
  box7: number;
}

export interface BoxDistributionResponse {
  wordlistId: number;
  distribution: BoxDistribution;
  totalWords: number;
}

export interface HistoricalBoxDistribution {
  date: string;
  boxes: BoxDistribution;
}

export interface HistoricalBoxDistributionResponse {
  wordlistId: number;
  days: number;
  distribution: HistoricalBoxDistribution[];
}

export interface PracticeTimeStats {
  date: string;
  practiceTimeMs: number;
  practiceTimeMinutes: number;
  quizCount: number;
}

export interface PracticeTimeResponse {
  wordlistId: number;
  days: number;
  practiceTime: PracticeTimeStats[];
}

const BASE_URL = getApiBaseUrl();

// 1) Get statistics for a specific wordlist
export async function getWordlistStats(
  wordlistId: number,
): Promise<WordlistStats> {
  const endpoint = `${BASE_URL}/analytics/wordlists/${wordlistId}/stats`;
  const body = await callAPI<{ stats: WordlistStats }>("GET", endpoint);

  // Ensure valid stats structure with safe defaults
  const stats = body.stats || {};
  return {
    totalWords: stats.totalWords || 0,
    wordsMastered: stats.wordsMastered || 0,
    averageMastery: stats.averageMastery || 0,
    bestStreak: stats.bestStreak || 0,
    currentStreak: stats.currentStreak || 0,
    wordsStudiedToday: stats.wordsStudiedToday || 0,
    quizzesToday: stats.quizzesToday || 0,
    accuracyToday: stats.accuracyToday || 0,
  };
}

// 2) Get word mastery for a specific wordlist
export async function getWordlistWordMastery(
  wordlistId: number,
): Promise<WordMasteryStats[]> {
  const endpoint = `${BASE_URL}/analytics/wordlists/${wordlistId}/word-mastery`;
  return getAllPages<{ stats: WordMasteryStats[] }, WordMasteryStats>({
    endpoint,
    getItems: (payload) => requireArray(payload?.stats, "word mastery stats"),
    getItemKey: (stat) => requireItemKey(stat.wordId, "word mastery stat"),
  });
}

// 3) Get learning progress for the last N days
export async function getWordlistLearningProgress(
  wordlistId: number,
  days: number,
): Promise<LearningProgress[]> {
  const endpoint = `${BASE_URL}/analytics/wordlists/${wordlistId}/learning-progress?days=${days}`;
  const payload = await callAPI<{ progress: LearningProgress[] }>(
    "GET",
    endpoint,
  );
  return payload.progress || [];
}

// 4) Get quiz performance across all quiz types for a specific wordlist
export async function getWordlistQuizTypePerformance(
  wordlistId: number,
): Promise<QuizTypePerformance[]> {
  const endpoint = `${BASE_URL}/analytics/wordlists/${wordlistId}/quiz-type-performance`;
  const payload = await callAPI<{ quizPerformance: QuizTypePerformance[] }>(
    "GET",
    endpoint,
  );
  return payload.quizPerformance || [];
}

// 5) Get wordlist progress summary with mastery-based calculations
export interface WordlistProgressSummary {
  wordlistId: number;
  totalWords: number;
  wordsMastered: number;
  averageMastery: number;
  progressPercentage: number;
}

export async function getWordlistProgressSummary(
  wordlistId: number,
): Promise<WordlistProgressSummary> {
  const wordMasteryStats = await getWordlistWordMastery(wordlistId);
  return calculateWordlistProgressFromMastery(wordlistId, wordMasteryStats);
}

// Helper function to calculate progress summary from pre-fetched mastery data
export function calculateWordlistProgressFromMastery(
  wordlistId: number,
  wordMasteryStats: WordMasteryStats[],
): WordlistProgressSummary {
  // Safety check: ensure wordMasteryStats is valid array
  const safeStats = Array.isArray(wordMasteryStats) ? wordMasteryStats : [];

  const totalWords = safeStats.length;
  const wordsMastered = safeStats.filter(
    (word) =>
      word && typeof word.masteryLevel === "number" && word.masteryLevel >= 0.8, // Consider 80%+ as mastered
  ).length;

  const averageMastery =
    totalWords > 0
      ? safeStats.reduce((sum, word) => sum + (word?.masteryLevel || 0), 0) /
        totalWords
      : 0;

  const progressPercentage =
    totalWords > 0 ? Math.round(averageMastery * 100) : 0;

  return {
    wordlistId: wordlistId,
    totalWords: totalWords,
    wordsMastered: wordsMastered,
    averageMastery: averageMastery,
    progressPercentage: progressPercentage,
  };
}

// 6) Get current box distribution for a wordlist
export async function getWordlistCurrentBoxDistribution(
  wordlistId: number,
): Promise<BoxDistributionResponse> {
  const endpoint = `${BASE_URL}/analytics/wordlists/${wordlistId}/current-box-distribution`;
  const response = await callAPI<BoxDistributionResponse>("GET", endpoint);

  // Ensure valid response structure
  return {
    wordlistId: response.wordlistId || wordlistId,
    distribution: response.distribution || {
      box1: 0,
      box2: 0,
      box3: 0,
      box4: 0,
      box5: 0,
      box6: 0,
      box7: 0,
    },
    totalWords: response.totalWords || 0,
  };
}

// 7) Get historical box distribution for a wordlist
export async function getWordlistBoxDistributionHistory(
  wordlistId: number,
  days: number = 30,
): Promise<HistoricalBoxDistributionResponse> {
  const endpoint = `${BASE_URL}/analytics/wordlists/${wordlistId}/box-distribution-history?days=${days}`;
  const response = await callAPI<HistoricalBoxDistributionResponse>(
    "GET",
    endpoint,
  );

  // Ensure valid response structure
  return {
    wordlistId: response.wordlistId || wordlistId,
    days: response.days || days,
    distribution: Array.isArray(response.distribution)
      ? response.distribution
      : [],
  };
}

// 8) Get practice time for a wordlist
export async function getWordlistPracticeTime(
  wordlistId: number,
  days: number = 7,
): Promise<PracticeTimeResponse> {
  const endpoint = `${BASE_URL}/analytics/wordlists/${wordlistId}/practice-time?days=${days}`;
  const response = await callAPI<PracticeTimeResponse>("GET", endpoint);

  // Ensure valid response structure
  return {
    wordlistId: response.wordlistId || wordlistId,
    days: response.days || days,
    practiceTime: Array.isArray(response.practiceTime)
      ? response.practiceTime
      : [],
  };
}

// 9) Get progress summary for all wordlists - NEW BATCH ENDPOINT
export interface WordlistProgress {
  wordlistId: number;
  wordlistName: string;
  languageCode: string;
  totalWords: number;
  wordsMastered: number;
  progressPercent: number;
  dueCount: number;
  currentStreak: number;
  lastActivityDate?: string;
}

export interface ProgressSummaryResponse {
  wordlists: WordlistProgress[];
}

export async function getProgressSummary(): Promise<ProgressSummaryResponse> {
  const endpoint = `${BASE_URL}/analytics/progress-summary`;
  return {
    wordlists: await getAllPages<ProgressSummaryResponse, WordlistProgress>({
      endpoint,
      getItems: (response) =>
        requireArray(response?.wordlists, "progress summary wordlists"),
      getItemKey: (wordlist) =>
        requireItemKey(wordlist.wordlistId, "progress summary wordlist"),
    }),
  };
}

function requireArray<T>(value: unknown, responseField: string): T[] {
  if (!Array.isArray(value)) {
    throw new PaginationError(
      `Invalid paginated ${responseField} response: expected an array`,
    );
  }
  return value as T[];
}

function requireItemKey(value: unknown, itemName: string): string {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value <= 0) {
    throw new PaginationError(
      `Invalid paginated ${itemName} response: expected a positive ID`,
    );
  }
  return String(value);
}
